# Data Exploration

### Import libraries

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

### Load & combine data

In [2]:
df = pd.concat([pd.read_parquet(f'../../data/bahn_data/data-2025-{i:02}.parquet') for i in range(1, 5)], ignore_index=True)

### Display data shape

In [3]:
df.shape

(7748819, 13)

### Display first 5 rows

In [7]:
df.head()

,station,train_name,final_destination_station,delay_in_min,time,is_canceled,train_type,train_line_ride_id,train_line_station_num,arrival_planned_time,arrival_change_time,departure_planned_time,departure_change_time
0,Augsburg Hbf,NJ,Hamburg-Altona,45,2025-01-01 00:10:00,False,NJ,-8677182982400525824-2412312046,7,2024-12-31 23:23:00,2024-12-31 23:46:00,2024-12-31 23:25:00,2025-01-01 00:10:00
1,Augsburg Hbf,NJ,Amsterdam Centraal,45,2025-01-01 00:10:00,False,NJ,3272146161175325165-2412312046,7,2024-12-31 23:23:00,2024-12-31 23:46:00,2024-12-31 23:25:00,2025-01-01 00:10:00
2,Bielefeld Hbf,ICE 102,Hannover Hbf,48,2025-01-01 00:17:00,False,ICE,-9170319321262747816-2412311713,16,2024-12-31 23:28:00,2025-01-01 00:16:00,2024-12-31 23:29:00,2025-01-01 00:17:00
3,Bietigheim-Bissingen,MEX 18,Heilbronn Hbf,0,2025-01-01 00:00:00,False,MEX,-5488879328755233091-2412312233,13,2024-12-31 23:58:00,2024-12-31 23:59:00,2025-01-01 00:00:00,2025-01-01 00:00:00
4,Bietigheim-Bissingen,MEX 17a,Bietigheim-Bissingen,1,2025-01-01 00:00:00,False,MEX,-2703871813082427678-2412312323,12,2024-12-31 23:59:00,2025-01-01 00:00:00,NaT,NaT


In [8]:
df.dtypes

station                      string[python]
train_name                   string[python]
final_destination_station    string[python]
delay_in_min                          int32
time                         datetime64[ns]
is_canceled                         boolean
train_type                   string[python]
train_line_ride_id           string[python]
train_line_station_num                int32
arrival_planned_time         datetime64[ns]
arrival_change_time          datetime64[ns]
departure_planned_time       datetime64[ns]
departure_change_time        datetime64[ns]
dtype: object

### Create time columns for better analysis

In [9]:
# Create time columns
df["month"] = pd.to_datetime(df["time"]).dt.month
# rename month values to actual names
df["month"] = df["month"].replace({
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December"
})

df["weekday"] = pd.to_datetime(df["time"]).dt.weekday
# rename weekday values to actual names
df["weekday"] = df["weekday"].replace({
    0: "Monday",
    1: "Tuesday",
    2: "Wednesday",
    3: "Thursday",
    4: "Friday",
    5: "Saturday",
    6: "Sunday"
})
df["day"] = pd.to_datetime(df["time"]).dt.day
df["hour"] = pd.to_datetime(df["time"]).dt.hour
df["minute"] = pd.to_datetime(df["time"]).dt.minute

### Display one specific ride

In [10]:
df[df['train_line_ride_id'] == "-9170319321262747816-2501011713"].head(20)

,station,train_name,final_destination_station,delay_in_min,time,is_canceled,train_type,train_line_ride_id,train_line_station_num,arrival_planned_time,arrival_change_time,departure_planned_time,departure_change_time,month,weekday,day,hour,minute
1994,Bielefeld Hbf,ICE 102,Hannover Hbf,15,2025-01-01 23:33:00,False,ICE,-9170319321262747816-2501011713,16,2025-01-01 23:17:00,2025-01-01 23:32:00,2025-01-01 23:18:00,2025-01-01 23:33:00,January,Wednesday,1,23,33
5748,Dortmund Hbf,ICE 102,Hannover Hbf,18,2025-01-01 22:50:00,False,ICE,-9170319321262747816-2501011713,13,2025-01-01 22:30:00,2025-01-01 22:48:00,2025-01-01 22:32:00,2025-01-01 22:50:00,January,Wednesday,1,22,50
6744,Düsseldorf Hbf,ICE 102,Hannover Hbf,1,2025-01-01 21:35:00,False,ICE,-9170319321262747816-2501011713,10,2025-01-01 21:32:00,2025-01-01 21:32:00,2025-01-01 21:34:00,2025-01-01 21:35:00,January,Wednesday,1,21,35
7541,Duisburg Hbf,ICE 102,Hannover Hbf,6,2025-01-01 21:53:00,False,ICE,-9170319321262747816-2501011713,11,2025-01-01 21:45:00,2025-01-01 21:50:00,2025-01-01 21:47:00,2025-01-01 21:53:00,January,Wednesday,1,21,53
9522,Essen Hbf,ICE 102,Hannover Hbf,7,2025-01-01 22:07:00,False,ICE,-9170319321262747816-2501011713,12,2025-01-01 21:58:00,2025-01-01 22:03:00,2025-01-01 22:00:00,2025-01-01 22:07:00,January,Wednesday,1,22,7
10929,Freiburg(Breisgau) Hbf,ICE 102,Hannover Hbf,5,2025-01-01 18:00:00,False,ICE,-9170319321262747816-2501011713,3,2025-01-01 17:53:00,2025-01-01 17:58:00,2025-01-01 17:55:00,2025-01-01 18:00:00,January,Wednesday,1,18,0
14347,Hamm(Westf)Hbf,ICE 102,Hannover Hbf,18,2025-01-01 23:06:00,False,ICE,-9170319321262747816-2501011713,14,2025-01-01 22:46:00,2025-01-01 23:04:00,2025-01-01 22:48:00,2025-01-01 23:06:00,January,Wednesday,1,23,6
16253,Herford,ICE 102,Hannover Hbf,13,2025-01-01 23:41:00,False,ICE,-9170319321262747816-2501011713,17,2025-01-01 23:26:00,2025-01-01 23:40:00,2025-01-01 23:28:00,2025-01-01 23:41:00,January,Wednesday,1,23,41
18294,Karlsruhe Hbf,ICE 102,Hannover Hbf,3,2025-01-01 19:04:00,False,ICE,-9170319321262747816-2501011713,6,2025-01-01 18:58:00,2025-01-01 19:02:00,2025-01-01 19:01:00,2025-01-01 19:04:00,January,Wednesday,1,19,4
20334,Köln Hbf,ICE 102,Hannover Hbf,1,2025-01-01 21:12:00,False,ICE,-9170319321262747816-2501011713,9,2025-01-01 21:05:00,2025-01-01 21:05:00,2025-01-01 21:11:00,2025-01-01 21:12:00,January,Wednesday,1,21,12


In [12]:
df.tail()

,station,train_name,final_destination_station,delay_in_min,time,is_canceled,train_type,train_line_ride_id,train_line_station_num,arrival_planned_time,arrival_change_time,departure_planned_time,departure_change_time,month,weekday,day,hour,minute,holiday
7748814,Hamburg-Altona,S 1,Hamburg-Blankenese,0,2025-04-30 23:21:00,False,S,1196912460244153290-2504302242,17,2025-04-30 23:20:00,2025-04-30 23:20:00,2025-04-30 23:21:00,2025-04-30 23:21:00,April,Wednesday,30,23,21,0
7748815,Hamburg-Altona,S 1,Wedel(Holst),0,2025-04-30 23:51:00,False,S,2348175559750561464-2504302312,17,2025-04-30 23:50:00,2025-04-30 23:50:00,2025-04-30 23:51:00,2025-04-30 23:51:00,April,Wednesday,30,23,51,0
7748816,Hamburg-Altona,S 2,Aumühle,0,2025-04-30 23:32:00,False,S,2422774709714075071-2504302332,1,NaT,NaT,2025-04-30 23:32:00,2025-04-30 23:32:00,April,Wednesday,30,23,32,0
7748817,Hamburg-Altona,S 1,Wedel(Holst),0,2025-04-30 23:31:00,False,S,8805686558806681772-2504302248,20,2025-04-30 23:30:00,2025-04-30 23:30:00,2025-04-30 23:31:00,2025-04-30 23:31:00,April,Wednesday,30,23,31,0
7748818,Hamburg-Altona,S 2,Aumühle,0,2025-04-30 23:52:00,False,S,-8259733845561586295-2504302352,1,NaT,NaT,2025-04-30 23:52:00,2025-04-30 23:52:00,April,Wednesday,30,23,52,0


### Search for cancellations in the data

In [5]:
# Display all rows with where is_canceled is True and train type is ICE
df[(df['is_canceled'] == True) & (df['train_type'] == 'ICE')].head(20)

,station,train_name,final_destination_station,delay_in_min,time,is_canceled,train_type,train_line_ride_id,train_line_station_num,arrival_planned_time,arrival_change_time,departure_planned_time,departure_change_time
575,Aschaffenburg Hbf,ICE 228,Dortmund Hbf,42,2025-01-01 13:46:00,True,ICE,-3893113454994955051-2501010713,10,2025-01-01 13:03:00,2025-01-01 13:44:00,2025-01-01 13:04:00,2025-01-01 13:46:00
646,Aschaffenburg Hbf,ICE 1222,Dortmund Hbf,6,2025-01-01 19:16:00,True,ICE,4218053850205204069-2501011618,5,2025-01-01 19:08:00,2025-01-01 19:14:00,2025-01-01 19:10:00,2025-01-01 19:16:00
790,Augsburg Hbf,ICE 1694,Berlin Hbf,35,2025-01-01 09:48:00,True,ICE,2342573116787471412-2501010838,3,2025-01-01 09:11:00,2025-01-01 09:46:00,2025-01-01 09:13:00,2025-01-01 09:48:00
1067,Augsburg Hbf,ICE 1205,München Hbf,0,2025-01-01 19:38:00,True,ICE,-1151057268452720658-2501011311,15,2025-01-01 19:36:00,2025-01-01 19:36:00,2025-01-01 19:38:00,2025-01-01 19:38:00
1104,Augsburg Hbf,ICE 1099,München Hbf,0,2025-01-01 21:41:00,True,ICE,-8959018065734509300-2501011818,9,2025-01-01 21:40:00,2025-01-01 21:40:00,2025-01-01 21:41:00,2025-01-01 21:41:00
1158,Augsburg Hbf,ICE 615,München Hbf,46,2025-01-02 00:41:00,True,ICE,-5425302918995479956-2501011529,17,2025-01-01 23:53:00,2025-01-02 00:39:00,2025-01-01 23:55:00,2025-01-02 00:41:00
2832,Bonn Hbf,ICE 929,Passau Hbf,0,2025-01-01 18:15:00,True,ICE,8428028954903351107-2501011636,7,2025-01-01 18:13:00,2025-01-01 18:13:00,2025-01-01 18:15:00,2025-01-01 18:15:00
3553,Bremen Hbf,ICE 630,Bremen Hbf,1,2025-01-01 14:46:00,True,ICE,8240768927220221024-2501010917,9,2025-01-01 14:45:00,2025-01-01 14:46:00,NaT,NaT
3966,Bruchsal,ICE 1578,Greifswald,0,2025-01-01 11:22:00,True,ICE,-2849772715769172580-2501011110,2,2025-01-01 11:21:00,2025-01-01 11:21:00,2025-01-01 11:22:00,2025-01-01 11:22:00
4268,Darmstadt Hbf,ICE 934,Hamburg-Altona,-3,2025-01-01 06:34:00,True,ICE,-5096032800673842604-2501010637,1,NaT,NaT,2025-01-01 06:37:00,2025-01-01 06:34:00


### Display specific ride id to check where its canceled and then confirm with subtrips dataframe

In [12]:
df[(df['train_line_ride_id'] == "4218053850205204069-2501011618")]

,station,train_name,final_destination_station,delay_in_min,time,is_canceled,train_type,train_line_ride_id,train_line_station_num,arrival_planned_time,arrival_change_time,departure_planned_time,departure_change_time
646,Aschaffenburg Hbf,ICE 1222,Dortmund Hbf,6,2025-01-01 19:16:00,True,ICE,4218053850205204069-2501011618,5,2025-01-01 19:08:00,2025-01-01 19:14:00,2025-01-01 19:10:00,2025-01-01 19:16:00
5789,Dortmund Hbf,ICE 1222,Dortmund Hbf,22,2025-01-02 00:04:00,False,ICE,4218053850205204069-2501011618,14,2025-01-01 23:42:00,2025-01-02 00:04:00,NaT,NaT
7895,Solingen Hbf,ICE 1222,Dortmund Hbf,21,2025-01-01 23:14:00,False,ICE,4218053850205204069-2501011618,11,2025-01-01 22:52:00,2025-01-01 23:12:00,2025-01-01 22:53:00,2025-01-01 23:14:00
10460,Frankfurt(Main)Hbf,ICE 1222,Dortmund Hbf,23,2025-01-01 20:22:00,False,ICE,4218053850205204069-2501011618,6,2025-01-01 19:52:00,2025-01-01 20:14:00,2025-01-01 19:59:00,2025-01-01 20:22:00
13631,Hagen Hbf,ICE 1222,Dortmund Hbf,21,2025-01-01 23:44:00,False,ICE,4218053850205204069-2501011618,13,2025-01-01 23:21:00,2025-01-01 23:42:00,2025-01-01 23:23:00,2025-01-01 23:44:00
17436,Ingolstadt Hbf,ICE 1222,Dortmund Hbf,0,2025-01-01 16:59:00,False,ICE,4218053850205204069-2501011618,2,2025-01-01 16:57:00,2025-01-01 16:58:00,2025-01-01 16:59:00,2025-01-01 16:59:00
19322,Koblenz Hbf,ICE 1222,Dortmund Hbf,24,2025-01-01 21:54:00,False,ICE,4218053850205204069-2501011618,9,2025-01-01 21:28:00,2025-01-01 21:39:00,2025-01-01 21:30:00,2025-01-01 21:54:00
21744,Mainz Hbf,ICE 1222,Dortmund Hbf,16,2025-01-01 20:53:00,False,ICE,4218053850205204069-2501011618,8,2025-01-01 20:35:00,2025-01-01 20:51:00,2025-01-01 20:37:00,2025-01-01 20:53:00
23624,Würzburg Hbf,ICE 1222,Dortmund Hbf,15,2025-01-01 18:46:00,False,ICE,4218053850205204069-2501011618,4,2025-01-01 18:27:00,2025-01-01 18:44:00,2025-01-01 18:31:00,2025-01-01 18:46:00
24167,München Hbf,ICE 1222,Dortmund Hbf,0,2025-01-01 16:18:00,False,ICE,4218053850205204069-2501011618,1,NaT,NaT,2025-01-01 16:18:00,2025-01-01 16:18:00


# Explore subtrips dataframe after transforming the data

In [2]:
# Load subtrips data
df_subtrips = pd.read_parquet('../../data/bahn_data/processed/subtrips_data.parquet')

In [3]:
df_subtrips.shape

(4055884, 9)

### Display specific ride id (check if canceled is working properly)

In [4]:
df_subtrips[df_subtrips['ride_id'] == "4218053850205204069-2501011618"]

,ride_id,train_name,origin_station,destination_station,departure_time_origin,day_of_week,hour_of_day,delay_at_dest,canceled
2702662,4218053850205204069-2501011618,ICE 1222,München Hbf,Ingolstadt Hbf,2025-01-01 16:18:00,2,16,0,False
2702663,4218053850205204069-2501011618,ICE 1222,München Hbf,Nürnberg Hbf,2025-01-01 16:18:00,2,16,14,False
2702664,4218053850205204069-2501011618,ICE 1222,München Hbf,Würzburg Hbf,2025-01-01 16:18:00,2,16,15,False
2702665,4218053850205204069-2501011618,ICE 1222,München Hbf,Aschaffenburg Hbf,2025-01-01 16:18:00,2,16,6,True
2702666,4218053850205204069-2501011618,ICE 1222,München Hbf,Frankfurt(Main)Hbf,2025-01-01 16:18:00,2,16,23,False
...,...,...,...,...,...,...,...,...,...
2702735,4218053850205204069-2501011618,ICE 1222,Solingen Hbf,Hagen Hbf,2025-01-01 22:53:00,2,22,21,False
2702736,4218053850205204069-2501011618,ICE 1222,Solingen Hbf,Dortmund Hbf,2025-01-01 22:53:00,2,22,22,False
2702737,4218053850205204069-2501011618,ICE 1222,Wuppertal Hbf,Hagen Hbf,2025-01-01 23:06:00,2,23,21,False
2702738,4218053850205204069-2501011618,ICE 1222,Wuppertal Hbf,Dortmund Hbf,2025-01-01 23:06:00,2,23,22,False


As expected every subtrip with Aschaffenburg as origin and destination is canceled.

### Show all rows with ICE 102 as train_name

In [5]:
df_subtrips[df_subtrips['train_name'] == "ICE 102"].head(20)

,ride_id,train_name,origin_station,destination_station,departure_time_origin,day_of_week,hour_of_day,delay_at_dest,canceled
1936612,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Mannheim Hbf,2025-01-18 19:01:00,5,19,0,False
1936613,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Köln Hbf,2025-01-18 19:01:00,5,19,0,False
1936614,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Düsseldorf Hbf,2025-01-18 19:01:00,5,19,3,False
1936615,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Duisburg Hbf,2025-01-18 19:01:00,5,19,2,False
1936616,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Essen Hbf,2025-01-18 19:01:00,5,19,0,False
1936617,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Dortmund Hbf,2025-01-18 19:01:00,5,19,20,False
1936618,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Hamm(Westf)Hbf,2025-01-18 19:01:00,5,19,21,False
1936619,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Bielefeld Hbf,2025-01-18 19:01:00,5,19,17,False
1936620,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Herford,2025-01-18 19:01:00,5,19,19,False
1936621,-8957484325052121716-2501181901,ICE 102,Karlsruhe Hbf,Hannover Hbf,2025-01-18 19:01:00,5,19,14,False


# After merging the subtrips with the weather data, we can explore the data further.

In [14]:
# Read the merged data
df_merged = pd.read_parquet('../../data/subtrips_with_weather.parquet')

In [15]:
df_merged.shape

(4055884, 17)

In [16]:
df_merged.head(20)

,ride_id,train_name,station,destination_station,departure_time_origin,day_of_week,hour_of_day,delay_at_dest,canceled,time_hour,time,temp_celsius,humidity,wind_speed,weather,rain_amount,snow_amount
0,-1028270319216920413-2504121732,ICE 1695,Nürnberg Hbf,Augsburg Hbf,2025-04-12 17:32:00,5,17,8,False,2025-04-12 17:00:00,2025-04-12 17:00:00,21.11,32.0,3.60,clear sky,0.00,0.0
1,-1028270319216920413-2504121732,ICE 1695,Nürnberg Hbf,München Hbf,2025-04-12 17:32:00,5,17,12,False,2025-04-12 17:00:00,2025-04-12 17:00:00,21.11,32.0,3.60,clear sky,0.00,0.0
2,-1028270319216920413-2504121732,ICE 1695,Augsburg Hbf,München Hbf,2025-04-12 18:48:00,5,18,12,False,2025-04-12 18:00:00,2025-04-12 18:00:00,19.83,34.0,2.57,clear sky,0.00,0.0
3,-1028270319216920413-2504131732,ICE 1695,Nürnberg Hbf,Augsburg Hbf,2025-04-13 17:32:00,6,17,2,False,2025-04-13 17:00:00,2025-04-13 17:00:00,18.85,55.0,2.57,scattered clouds,0.00,0.0
4,-1028270319216920413-2504131732,ICE 1695,Nürnberg Hbf,München Hbf,2025-04-13 17:32:00,6,17,-1,False,2025-04-13 17:00:00,2025-04-13 17:00:00,18.85,55.0,2.57,scattered clouds,0.00,0.0
5,-1028270319216920413-2504131732,ICE 1695,Augsburg Hbf,München Hbf,2025-04-13 18:48:00,6,18,-1,False,2025-04-13 18:00:00,2025-04-13 18:00:00,16.82,59.0,3.60,clear sky,0.00,0.0
6,-1028270319216920413-2504141732,ICE 1695,Nürnberg Hbf,Augsburg Hbf,2025-04-14 17:32:00,0,17,0,True,2025-04-14 17:00:00,2025-04-14 17:00:00,19.00,55.0,1.54,few clouds,0.00,0.0
7,-1028270319216920413-2504141732,ICE 1695,Nürnberg Hbf,München Hbf,2025-04-14 17:32:00,0,17,0,True,2025-04-14 17:00:00,2025-04-14 17:00:00,19.00,55.0,1.54,few clouds,0.00,0.0
8,-1028270319216920413-2504141732,ICE 1695,Augsburg Hbf,München Hbf,2025-04-14 18:48:00,0,18,0,True,2025-04-14 18:00:00,2025-04-14 18:00:00,17.91,59.0,3.09,few clouds,0.00,0.0
9,-1028270319216920413-2504151732,ICE 1695,Nürnberg Hbf,Augsburg Hbf,2025-04-15 17:32:00,1,17,20,False,2025-04-15 17:00:00,2025-04-15 17:00:00,17.01,68.0,5.14,light intensity shower rain,3.16,0.0


In [17]:
# Display -9170319321262747816-2501011713 ride_id
df_merged[df_merged['ride_id'] == "-9170319321262747816-2501011713"].head(20)

,ride_id,train_name,station,destination_station,departure_time_origin,day_of_week,hour_of_day,delay_at_dest,canceled,time_hour,time,temp_celsius,humidity,wind_speed,weather,rain_amount,snow_amount
1976720,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Offenburg,2025-01-01 17:55:00,2,17,4,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976721,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Karlsruhe Hbf,2025-01-01 17:55:00,2,17,3,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976722,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Mannheim Hbf,2025-01-01 17:55:00,2,17,1,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976723,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Köln Hbf,2025-01-01 17:55:00,2,17,1,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976724,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Düsseldorf Hbf,2025-01-01 17:55:00,2,17,1,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976725,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Duisburg Hbf,2025-01-01 17:55:00,2,17,6,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976726,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Essen Hbf,2025-01-01 17:55:00,2,17,7,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976727,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Dortmund Hbf,2025-01-01 17:55:00,2,17,18,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976728,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Hamm(Westf)Hbf,2025-01-01 17:55:00,2,17,18,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0
1976729,-9170319321262747816-2501011713,ICE 102,Freiburg(Breisgau) Hbf,Bielefeld Hbf,2025-01-01 17:55:00,2,17,15,False,2025-01-01 17:00:00,2025-01-01 17:00:00,6.31,60.0,2.68,scattered clouds,0.0,0.0


### There were a lot of NaN because of windows being stupid and not using uft-8 encoding when reading json

Now its better after changing the encoding to utf-8

In [18]:
# Check for NaN values in the merged data
df_merged.isna().sum()

ride_id                      0
train_name                   0
station                      0
destination_station          0
departure_time_origin        0
day_of_week                  0
hour_of_day                  0
delay_at_dest                0
canceled                     0
time_hour                    0
time                     10445
temp_celsius             10445
humidity                 10445
wind_speed               10445
weather                  10445
rain_amount              10445
snow_amount              10445
dtype: int64

In [19]:
# Display all rows with NaN values
df_merged[df_merged.isna().any(axis=1)].head(20)

,ride_id,train_name,station,destination_station,departure_time_origin,day_of_week,hour_of_day,delay_at_dest,canceled,time_hour,time,temp_celsius,humidity,wind_speed,weather,rain_amount,snow_amount
591,-1049617350581839421-2501060637,IC 2447,Hannover Hbf,Braunschweig Hbf,2025-01-06 06:37:00,0,6,1,False,2025-01-06 06:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
592,-1049617350581839421-2501060637,IC 2447,Hannover Hbf,Magdeburg Hbf,2025-01-06 06:37:00,0,6,9,False,2025-01-06 06:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
593,-1049617350581839421-2501060637,IC 2447,Hannover Hbf,Halle(Saale)Hbf,2025-01-06 06:37:00,0,6,5,False,2025-01-06 06:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
594,-1049617350581839421-2501060637,IC 2447,Hannover Hbf,Leipzig Hbf,2025-01-06 06:37:00,0,6,1,False,2025-01-06 06:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
595,-1049617350581839421-2501060637,IC 2447,Hannover Hbf,Dresden-Neustadt,2025-01-06 06:37:00,0,6,26,False,2025-01-06 06:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
596,-1049617350581839421-2501060637,IC 2447,Hannover Hbf,Dresden Hbf,2025-01-06 06:37:00,0,6,31,False,2025-01-06 06:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
597,-1049617350581839421-2501060637,IC 2447,Braunschweig Hbf,Magdeburg Hbf,2025-01-06 07:10:00,0,7,9,False,2025-01-06 07:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
598,-1049617350581839421-2501060637,IC 2447,Braunschweig Hbf,Halle(Saale)Hbf,2025-01-06 07:10:00,0,7,5,False,2025-01-06 07:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
599,-1049617350581839421-2501060637,IC 2447,Braunschweig Hbf,Leipzig Hbf,2025-01-06 07:10:00,0,7,1,False,2025-01-06 07:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN
600,-1049617350581839421-2501060637,IC 2447,Braunschweig Hbf,Dresden-Neustadt,2025-01-06 07:10:00,0,7,26,False,2025-01-06 07:00:00,NaT,NaN,NaN,NaN,None,NaN,NaN


It seems that the weather data is not available for all rides. But we still have over 2.7 Mio rides with weather data, which is a good amount for analysis and model training.

After adding IC to the data, we now have a bigger dataset with more than 4.05 Mio rides.